inference-engine/src/mkldnn_plugin/nodes/mkldnn_topk_node.cpp

```diff
@@ -40,190 +1041,541 @@ bool MKLDNNTopKNode::isSupportedOperation(const std::shared_ptr<ngraph::Node>& o
+                pushDesc(memory::format_tag::ndhwc, memory::format_tag::ndhwc, data_type, impl_type);
+                pushDesc(memory::format_tag::nCdhw8c, memory::format_tag::nCdhw8c, data_type, impl_type);
+            }
+        }
+    } else {
+        pushDesc(MKLDNNMemory::GetPlainFormat(memory::dims(getParentEdgeAt(TOPK_DATA)->getDims().ndims())),
+             MKLDNNMemory::GetPlainFormat(memory::dims(getChildEdgeAt(TOPK_DATA)->getDims().ndims())), memory::data_type::f32, impl_desc_type::ref);
+    }
+}
+
+void MKLDNNTopKNode::createPrimitive() {
+    auto &dstMemPtr = getChildEdgeAt(TOPK_DATA)->getMemoryPtr();
+    auto &srcDataMemPtr = getParentEdgeAt(TOPK_DATA)->getMemoryPtr();
+    if (!dstMemPtr || !dstMemPtr->GetPrimitivePtr())
+        IE_THROW() << errorPrefix << " has not allocated destination memory.";
+    if (!srcDataMemPtr || !srcDataMemPtr->GetPrimitivePtr())
+        IE_THROW() << errorPrefix << " has not allocate input memory.";
+    if (getSelectedPrimitiveDescriptor() == nullptr)
+        IE_THROW() << errorPrefix << " has nullable preferable primitive descriptor";
+
+    auto selectedPD = getSelectedPrimitiveDescriptor();
+    Layout selected_layout = selectedPD->getConfig().inConfs[TOPK_DATA].desc.getLayout();
+    if (MKLDNNMemory::GetPlainLayout(getParentEdgeAt(TOPK_DATA)->getDims()) == selected_layout) {
+        layout = TopKLayoutType::topk_ncsp;
+    } else if ((selected_layout == NHWC) || (selected_layout == NDHWC)) {
+        layout = TopKLayoutType::topk_nspc;
+    } else {
+        layout = TopKLayoutType::topk_blocked;
+    }
+
+    topk_innermost = false;
+    if ((layout == TopKLayoutType::topk_ncsp && axis == dst_dims_size - 1) ||
+       ((layout == TopKLayoutType::topk_nspc || layout == TopKLayoutType::topk_blocked) && axis == 1))
+        topk_innermost = true;
+
+    if (mayiuse(cpu::x64::avx512_common)) {
+        blk_size = 16;
+        count_xmm = 16; // only 16 vector registers are valid in sse instructions
+    } else if (mayiuse(cpu::x64::sse41)) {
+        blk_size = 8;
+        count_xmm = 16;
+    }
+
+    // [case 1]: if 2 * (top_k + 1) + 2 <= count_vec, thus top_k is relatively small, and vector registers are sufficient
+    //           to keep all necessary data for sorting, no need to load and store frequently, use bubble sort;
+    // [case 2]: otherwise, only when topk is imposed on innermost dimsension of planar(ncsp/nspc) layout, should heap sort be used;
+    // [case 3]: bitonic sort is used as default
+    if (top_k <= count_xmm / 2 - 2) {
+        algorithm = TopKAlgorithm::topk_bubble_sort;
+    } else if ((layout == TopKLayoutType::topk_ncsp || layout == TopKLayoutType::topk_nspc) && topk_innermost) {
+        algorithm = TopKAlgorithm::topk_heap_sort;
+    } else {
+        algorithm = TopKAlgorithm::topk_bitonic_sort;
+    }
+
+    if (jit_mode) {
+        O = 1, I = 1;
+        A = src_dims[axis];
+        if (layout == TopKLayoutType::topk_ncsp) {
```

Comment:
Please provide the comment describing purpose of the code block below. Maybe it makes sense to extract it in separate function with meaningful name.

Reply:
Fixed. Thanks Dmitry!